node/chainSimulator/process/processor.go

```diff
@@ -106,17 +108,97 @@ func (creator *blocksCreator) CreateNewBlock() error {
 		return err
 	}
 
+	notNil := !check.IfNil(currentHeader)
+	enableEpochHandler := creator.nodeHandler.GetCoreComponents().EnableEpochsHandler()
+	var previousProof *dataBlock.HeaderProof
+	if notNil && enableEpochHandler.IsFlagEnabled(common.EquivalentMessagesFlag) {
+		sig, errS := creator.generateSignature(prevHash, blsKey.PubKey(), currentHeader)
+		if errS != nil {
+			return errS
+		}
+
+		previousProof = &dataBlock.HeaderProof{
+			PubKeysBitmap:       pubKeyBitmap,
+			AggregatedSignature: sig,
+			HeaderHash:          prevHash,
+			HeaderEpoch:         currentHeader.GetEpoch(),
+			HeaderNonce:         currentHeader.GetNonce(),
+			HeaderShardId:       currentHeader.GetShardID(),
+			HeaderRound:         currentHeader.GetNonce(),
+			IsStartOfEpoch:      currentHeader.IsStartOfEpochBlock(),
+		}
+
+		wasAdded := creator.nodeHandler.GetDataComponents().Datapool().Proofs().AddProof(previousProof)
+		if !wasAdded {
+			log.Warn("proofsPool.AddProof not added", "currHdrHash", prevHash)
+		}
+	}
+
 	header, block, err := bp.CreateBlock(newHeader, func() bool {
 		return true
 	})
 	if err != nil {
 		return err
 	}
 
-	err = creator.setHeaderSignatures(header, blsKey.PubKey())
+	if notNil && common.ShouldBlockHavePrevProof(header, enableEpochHandler, common.EquivalentMessagesFlag) {
+		selectionEpoch := epoch
```

Comment:
here you could extract into a updatePrevProofAndAddOnHeader(header, currentHeader, prevProof) function

Reply:
refactored